### ISSUE TITLE
Pull in sprig functions?

### ISSUE THREAD

**Issue by docwhat** [hairyhenderson/gomplate#283]
[Sprig](http://masterminds.github.io/sprig/) looks like it has a lot of really cool functions.  What about ~stealing~ using their library of functions in gomplate?

**Comment by hairyhenderson** [comment]
@docwhat weird coincidence - I was just thinking about that yesterday 😂...

There's definitely some overlap, there's lots of things in Sprig that aren't yet in gomplate. Unfortunately the functions themselves aren't exported, so really we'd just be talking about copying the same functions into gomplate.

An option would be to use the FuncMap that Sprig exports, and simply add that into the chain, but I'm not a big fan of that since it "pollutes" the root namespace.

Basically, I think the right approach would be to be more methodical and bring in the specific functions that are missing, within the appropriate namespace.

Or if the Sprig maintainers could be convinced to export all of the functions 🤔...

**Comment by docwhat** [comment]
Can't you just walk `FuncMap` or `TxtFuncMap` and copy what you want into a new map with whatever name you want?  Everything is exported (indirectly) via the Map.

I didn't know if you knew that `.Funcs(someFuncMap)` can be called multiple times.  I just [discovered](https://github.com/docwhat/temple/commit/aebab900e46bc2c8e5b944d51c6ac45a69813f8d#diff-4b8d8686fe71a23ce9bf8e85babe078dR60) that.

PS: [`temple`](https://github.com/docwhat/temple) got Sprig functions before you. 😝

**Comment by hairyhenderson** [comment]
> PS: temple got Sprig functions before you. 😝

😂

**Comment by hairyhenderson** [comment]
> Can't you just walk FuncMap or TxtFuncMap and copy what you want into a new map with whatever name you want? Everything is exported (indirectly) via the Map.

That's a really good point - hadn't considered that! I'll take a look to see how elegant (or inelegant) this can be.
 
> I didn't know if you knew that .Funcs(someFuncMap) can be called multiple times. I just discovered that.

Yup!

**Comment by hairyhenderson** [comment]
> > Can't you just walk FuncMap or TxtFuncMap and copy what you want into a new map with whatever name you want? Everything is exported (indirectly) via the Map.

> ... I'll take a look to see how elegant (or inelegant) this can be.

So... Pretty _inelegant_, it turns out. With the exception of the (mostly legacy) functions in the root namespace, the only entries in gomplate's main FuncMap are namespaces like `strings`. There is no entry for `strings.Trim` for example, because function names can't include a `.`.

If I wanted to add the `abbrev` function to gomplate's `strings` namespace as `strings.Abbrev`, I'd have to do something like this:

```go
// Abbrev -
func (f *StringFuncs) Abbrev(width int, s string) (string, error) {
	a := sprig.GenericFuncMap()["abbrev"]
	switch v := a.(type) {
	case func(int, string) string:
		return v(width, s), nil
	default:
		panic("wat")
	}
}
```

TBH, this is more trouble than it's worth - especially (in this case) as there's actually an exported `util.Abbrev` function from https://github.com/Masterminds/goutils...

So I think instead of trying to shoehorn `sprig` into `gomplate`, I'll see about adding in equivalent functionality that makes sense.

Thanks for logging the issue though. This is a good source of ideas for new functions!

**Comment by djgilcrease** [comment]
would not be overly difficult to import the sprig functions as sprig_* not properly namespaced, but not as bad as how they are doing it by default

```go
package funcs

import (
	"github.com/Masterminds/sprig"
	"fmt"
)

func AddSprigFuncs(f map[string]interface{}) {
	for fn, fc := range sprig.TxtFuncMap() {
		f[fmt.Sprintf("sprig_%s", fn)] = fc
	}

}
```

**Comment by djgilcrease** [comment]
No, I was just commenting on the open task if someone absolutely needed it. I like your method of namespacing the template funcs, might be interesting to split it out into it's own project

**Comment by hairyhenderson** [comment]
Ok, I'm going to close this issue out then. If there's any specific new functions that anyone needs - please open a new issue 🙂

**Comment by hickey** [comment]
Not looking to reopen this issue, but the big Sprig function that would be **really** nice to have access to is `default`. This would allow default values to be specified or input values to be overridden by other input values. 

I am sure that there are probably a couple more that might be useful, but that is the biggie that I see.

**Comment by hairyhenderson** [comment]
@hickey have you seen https://docs.gomplate.ca/functions/conv/#conv-default?

(also please open a new issue if you're looking for new functions, or start a [discussion](https://github.com/hairyhenderson/gomplate/discussions))

**Comment by hickey** [comment]
Ah, no I have not seen that. I would not have thought that the function would have been classified as a conversion method. Thanks for the pointer.

**Comment by turboBasic** [comment]
I am missing `dig` function from Sprig.  I’m still getting to grips with Gomplate (even though I’m already using it in a real production project ☺️), so once I’m confident that there’s no equivalent function available, I’ll go ahead and open a new issue for it.
